### PR TITLE
Extend Unicode support

### PR DESCRIPTION
**Pull request by mrylov**

### CODE REVIEW

File: src/odbc/DatabaseMetaDataBase.h [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+#include <odbc/Forwards.h>
+#include <odbc/Types.h>
+//------------------------------------------------------------------------------
+namespace odbc {
+//------------------------------------------------------------------------------
+/**
+ * Provides information about the database.
+ *
+ * A DatabaseMetaData object can be created using
+ * Connection::getDatabaseMetadata().
```
Comment (stefanuhrig) [inline src/odbc/DatabaseMetaDataBase.h:15]: We should mention the unicode variant here as well.

File: src/odbc/DatabaseMetaDataBase.h [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+     * Checks whether the database supports the given transaction isolation
+     * level or not.
+     *
+     * @return   Returns true if the given transaction isolation level is
+     *           supported, otherwise false.
+     */
+    bool supportsTransactionIsolation(TransactionIsolationLevel level);
+
+protected:
+    StatementRef createStatement() {  return parent_->createStatement();  }
```
Comment (stefanuhrig) [inline src/odbc/DatabaseMetaDataBase.h:89]: For consistency reasons I'd only declare `createStatement()` here and define it in the `cpp` file.

File: src/odbc/DatabaseMetaDataBase.cpp [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+u16string DatabaseMetaDataBase::getStringTypeInfoW(unsigned short typeInfo)
+{
+    vector<char16_t> buffer;
+    buffer.resize(256);
+    while (true)
+    {
+        SQLPOINTER ptr = &buffer[0];
+        SQLSMALLINT bufLen = (SQLSMALLINT)buffer.size();
+        SQLSMALLINT dataLen;
+        EXEC_DBC(SQLGetInfoW, parent_->hdbc_, typeInfo, ptr, bufLen, &dataLen);
```
Comment (stefanuhrig) [inline src/odbc/DatabaseMetaDataBase.cpp:113]: `SQLGetInfoW` expects the buffer size to be in bytes, not in characters.

File: src/odbc/DatabaseMetaDataBase.cpp [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        if (dataLen < bufLen)
+            break;
+        buffer.resize(dataLen + 1);
```
Comment (stefanuhrig) [inline src/odbc/DatabaseMetaDataBase.cpp:116]: We need to adapt that as well.

File: src/odbc/DatabaseMetaDataUnicode.h [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+#include <odbc/Forwards.h>
+#include <odbc/Types.h>
+//------------------------------------------------------------------------------
+namespace odbc {
+//------------------------------------------------------------------------------
+/**
+ * Provides information about the database.
+ *
+ * A DatabaseMetaData object can be created using
+ * Connection::getDatabaseMetadata().
```
Comment (stefanuhrig) [inline src/odbc/DatabaseMetaDataUnicode.h:16]: This comment needs adaption.

File: src/odbc/DatabaseMetaDataBase.h [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    std::string getStringTypeInfoA(unsigned short typeInfo);
+    std::u16string getStringTypeInfoW(unsigned short typeInfo);
```
Comment (stefanuhrig) [inline src/odbc/DatabaseMetaDataBase.h:91]: This method seems only to be used by the unicode class. I'd move it there.

File: src/odbc/PreparedStatement.h [hunk 7]
```diff
@@ -233,6 +233,14 @@ class ODBC_EXPORT PreparedStatement : public StatementBase
      */
     ResultSetMetaDataRef getMetaData();
 
+    /**
+     * Retrieves a ResultSetMetaData object that contains information about a
```
Comment (stefanuhrig) [inline src/odbc/PreparedStatement.h:237]: `ResultSetMetaDataUnicode`

File: src/odbc/PreparedStatement.h [hunk 8]
```diff
@@ -233,6 +233,14 @@ class ODBC_EXPORT PreparedStatement : public StatementBase
...
+     * ResultSet object returned by executeQuery method.
+     *
+     * @return  Returns a ResultSetMetaData object.
```
Comment (stefanuhrig) [inline src/odbc/PreparedStatement.h:240]: `ResultSetMetaDataUnicode`

File: src/odbc/ResultSet.h [hunk 9]
```diff
@@ -55,6 +57,14 @@ class ODBC_EXPORT ResultSet : public RefCounted
      */
     ResultSetMetaDataRef getMetaData();
 
+    /**
+     * Retrieves a ResultSetMetaData object that contains information about the
+     * number, types and properties of the columns in this ResultSet object.
+     *
+     * @return  Returns a ResultSetMetaData object.
```
Comment (stefanuhrig) [inline src/odbc/ResultSet.h:64]: `ResultSetMetaDataUnicode`

File: src/odbc/ResultSetMetaDataBase.h [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+     * @return             Returns true if the column can contain signed
+     *                     numbers, false otherwise.
+     */
+    bool isSigned(unsigned short columnIndex);
+
+protected:
+    std::string getStringColAttributeA(unsigned short columnIndex,
+        unsigned short field);
+    std::u16string getStringColAttributeW(unsigned short columnIndex,
+        unsigned short field);
```
Comment (stefanuhrig) [inline src/odbc/ResultSetMetaDataBase.h:155]: I'd move that to the Unicode class if it's only needed by it.

File: src/odbc/ResultSetMetaDataBase.h [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+     *
+     * @param columnIndex  The column index starting from 1.
+     * @return             Returns true if the column can contain signed
+     *                     numbers, false otherwise.
+     */
+    bool isSigned(unsigned short columnIndex);
+
+protected:
+    std::string getStringColAttributeA(unsigned short columnIndex,
+        unsigned short field);
```
Comment (stefanuhrig) [inline src/odbc/ResultSetMetaDataBase.h:153]: This method seems to be needed only in the non-Unicode class. I'd move it there.

File: src/odbc/ResultSetMetaDataBase.cpp [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+{
+    vector<char16_t> buffer;
+    buffer.resize(256);
+    while (true)
+    {
+        SQLPOINTER ptr = &buffer[0];
+        SQLSMALLINT bufLen = (SQLSMALLINT)buffer.size();
+        SQLSMALLINT dataLen;
+        EXEC_STMT(SQLColAttributeW, stmt_->hstmt_, columnIndex, field, ptr,
+            bufLen, &dataLen, NULL);
```
Comment (stefanuhrig) [inline src/odbc/ResultSetMetaDataBase.cpp:158]: The buffer size must be passed in bytes here.

File: src/odbc/ResultSetMetaDataBase.cpp [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+        if (dataLen < bufLen)
+            break;
+        buffer.resize(dataLen + 1);
```
Comment (stefanuhrig) [inline src/odbc/ResultSetMetaDataBase.cpp:161]: This must be adapted as well.

File: src/odbc/Connection.h [hunk 14]
```diff
@@ -309,6 +309,13 @@ class ODBC_EXPORT Connection : public RefCounted
      */
     DatabaseMetaDataRef getDatabaseMetaData();
 
+    /**
+     * Retrieves metadata information of the database.
+     *
+     * @return  Returns a reference to the DatabaseMetaData object.
```
[the remark this answers is not in the document]
Comment (mrylov) [reply src/odbc/Connection.h:315]: Done.

File: src/odbc/DatabaseMetaDataBase.h (hunk 1 above)
Comment (mrylov) [reply src/odbc/DatabaseMetaDataBase.h:15]: Done.

File: src/odbc/DatabaseMetaDataBase.h (hunk 2 above)
Comment (mrylov) [reply src/odbc/DatabaseMetaDataBase.h:89]: Done.

File: src/odbc/DatabaseMetaDataBase.cpp (hunk 3 above)
Comment (mrylov) [reply src/odbc/DatabaseMetaDataBase.cpp:113]: Done.

File: src/odbc/DatabaseMetaDataBase.cpp (hunk 4 above)
Comment (mrylov) [reply src/odbc/DatabaseMetaDataBase.cpp:116]: Done.

File: src/odbc/DatabaseMetaDataUnicode.h (hunk 5 above)
Comment (mrylov) [reply src/odbc/DatabaseMetaDataUnicode.h:16]: Done.

File: src/odbc/DatabaseMetaDataBase.h (hunk 6 above)
Comment (mrylov) [reply src/odbc/DatabaseMetaDataBase.h:91]: Done.

File: src/odbc/PreparedStatement.h (hunk 7 above)
Comment (mrylov) [reply src/odbc/PreparedStatement.h:237]: Done.

File: src/odbc/PreparedStatement.h (hunk 8 above)
Comment (mrylov) [reply src/odbc/PreparedStatement.h:240]: Done.

File: src/odbc/ResultSet.h (hunk 9 above)
Comment (mrylov) [reply src/odbc/ResultSet.h:64]: Done.

File: src/odbc/ResultSetMetaDataBase.h (hunk 10 above)
Comment (mrylov) [reply src/odbc/ResultSetMetaDataBase.h:155]: Done.

File: src/odbc/ResultSetMetaDataBase.cpp (hunk 12 above)
Comment (mrylov) [reply src/odbc/ResultSetMetaDataBase.cpp:158]: Done.

**Pull request merged**